### PR TITLE
slof_hugepage: Add slof_hugepage case for testing SLOF

### PR DESCRIPTION
**Pull request by YongxueHong**
slof_hugepage case is that verify SLOF info with hugepage.

ID: 1645009

Signed-off-by: Yongxue Hong <yhong@redhat.com>

### CODE REVIEW

**Comment by YongxueHong** [comment]
Depends on https://github.com/autotest/tp-qemu/pull/1531

**Comment by wangxianxian** [comment]
python3 ConfigTest.py --driveformat=virtio_scsi --nicmodel=virtio_net --platform=ppc64le --guestname=RHEL.x.x --vcpu=8 --mem=4096 --testcase=slof_hugepage --clone=no
slof_hugepage.with_backing_file_hotplug: PASS
slof_hugepage.without_backing_file_hotplug: PASS

**Comment by wangxianxian** [comment]
Acked by xianwang@redhat.com

File: qemu/tests/cfg/slof_hugepage.cfg [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            size_mem1 = 2G
+            policy_mem1 = default
+            mem-path_mem1 = /mnt/kvm_hugepage
+            plug_mem_name = plug
+            size_plug = 2G
+            policy_plug = bind
+            mem-path_plug = /mnt/kvm_hugepage
+            backend_plug = memory-backend-file
+            host-nodes_plug = 0
+            plug_timeout = 5
```
Comment (balamuruhans) [inline qemu/tests/cfg/slof_hugepage.cfg]: common params can be moved out of variants.

**Comment by YongxueHong** [comment]
@hereischen  @lolyu  help me review , thanks.

File: qemu/tests/slof_hugepage.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        4. mount this hugepage to /mnt/kvm_hugepage.
+    """
+    size = params['total_hugepage_size']
+    huge_page = test_setup.HugePageConfig(params)
+    error_context.context('Assign %sMB hugepages in host.' % size, logging.info)
+
+    hugepage_size = huge_page.get_hugepage_size()
+    logging.debug('Hugepage size is %skB in host.' % hugepage_size)
+
+    huge_page.target_hugepages = int((int(size) * 1024) / hugepage_size)
```
[the remark this answers is not in the document]
Comment (YongxueHong) [reply qemu/tests/slof_hugepage.py]: Thanks for your comment, updated.

File: qemu/tests/slof_hugepage.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    _setup_hugepage(params)
+
+    params['start_vm'] = 'yes'
+    env_process.preprocess_vm(test, params, env, params["main_vm"])
+
+    vm = env.get_vm(params["main_vm"])
+    content, next = slof.wait_for_loaded(vm, test)
```
[the remark this answers is not in the document]
Comment (YongxueHong) [reply qemu/tests/slof_hugepage.py]: Thanks for your comment, updated.

File: qemu/tests/slof_hugepage.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    content, _ = slof.wait_for_loaded(vm, test, next)
+
+    error_context.context("Check the output of SLOF.", logging.info)
+    slof.check_error(test, content)
+    logging.info("No any error info.")
+
+    error_context.context("Try to log into guest '%s'." % vm.name,
+                          logging.info)
+    session = vm.wait_for_login(timeout=timeout)
+    logging.info("log into guest '%s' successfully." % vm.name)
```
[the remark this answers is not in the document]
Comment (YongxueHong) [reply qemu/tests/slof_hugepage.py]: Thanks for your comment, have updated.

File: qemu/tests/slof_hugepage.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    huge_page = test_setup.HugePageConfig(params)
+    error_context.context('Assign %sMB hugepages in host.' % size, logging.info)
+
+    hugepage_size = huge_page.get_hugepage_size()
+    logging.debug('Hugepage size is %skB in host.' % hugepage_size)
+
+    huge_page.target_hugepages = int((int(size) * 1024) / hugepage_size)
+    logging.debug('Set hugepages to %d pages in host.'
+                  % huge_page.target_hugepages)
+    huge_page.set_hugepages()
```
[the remark this answers is not in the document]
Comment (YongxueHong) [reply qemu/tests/slof_hugepage.py]: Thanks for your comment, have updated.

File: qemu/tests/slof_hugepage.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    huge_page.mount_hugepage_fs()
+
+
+def _check_mem_increase(session, params, orig_mem):
+    """
+    check the size of memory increased in timeout.
+    :param org_mem:  int, original memory size.
+    """
+    end_time = time.time() + float(params.get('plug_timeout', 5))
+    while time.time() < end_time:
```
[the remark this answers is not in the document]
Comment (YongxueHong) [reply qemu/tests/slof_hugepage.py]: The utils_misc.wait_for() doesn't satisfy for my requirement.

File: qemu/tests/cfg/slof_hugepage.cfg (hunk 1 above)
Comment (YongxueHong) [reply qemu/tests/cfg/slof_hugepage.cfg]: Thanks for your comment, have updated.

File: qemu/tests/cfg/slof_hugepage.cfg [hunk 7]
```diff
@@ -0,0 +1,38 @@
+- slof_hugepage: install setup image_copy unattended_install.cdrom
+    virt_test_type = qemu
+    type = slof_hugepage
+    restart_vm = yes
+    kill_vm_on_error = yes
+    login_timeout = 240
```
Comment (lolyu) [inline qemu/tests/cfg/slof_hugepage.cfg]: `login_timeout` defaults to 240, the re-assignment is redundant.

File: qemu/tests/cfg/slof_hugepage.cfg [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    start_vm = no
+    slots_mem = 4
+    mem = 2048
+    maxmem_mem = 32G
+    pre_command = "echo 3 > /proc/sys/vm/drop_caches"
```
Comment (lolyu) [inline qemu/tests/cfg/slof_hugepage.cfg]: It is better to flush dirty pages first  before dropping caches:
`pre_command = "sync & echo 3 > /proc/sys/vm/drop_caches"`

Comment (lolyu) [reply qemu/tests/cfg/slof_hugepage.cfg]: Please replace commands with functions in `avocado.utils.memory`.

Comment (YongxueHong) [reply qemu/tests/cfg/slof_hugepage.cfg]: For the 2nd comment, what's the difference between with both of them?

File: qemu/tests/cfg/slof_hugepage.cfg (hunk 7 above)
Comment (YongxueHong) [reply qemu/tests/cfg/slof_hugepage.cfg]: thanks, have updated.

File: qemu/tests/slof_hugepage.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+                              logging.info)
+        timeout = float(params.get("login_timeout", 240))
+        session = vm.wait_for_login(timeout=timeout)
+        logging.info("log into guest '%s' successfully." % vm.name)
+        return session, next_pos
+
+    _setup_hugepage(params)
+
+    params['start_vm'] = 'yes'
+    env_process.preprocess_vm(test, params, env, params["main_vm"])
```
Comment (hereischen) [inline qemu/tests/slof_hugepage.py]: Do you need to add `not_preprocess = yes`? so it dose not `preprocess_vm` twice.

Comment (YongxueHong) [reply qemu/tests/slof_hugepage.py]: Thanks,  have updated.

**Comment by hereischen** [comment]
Acked-by: Haotong Chen hachen@redhat.com

**Pull request merged**